### PR TITLE
Shard coordinator remember entities refactor

### PR DESCRIPTION
**Pull request by johanandren**
References #28875 

Smaller scope than previous try.

### CODE REVIEW

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 1]
```diff
@@ -1105,36 +1106,51 @@ class DDataShardCoordinator(
...
-    if (rememberEntities) Set(CoordinatorStateKey, AllShardsKey) else Set(CoordinatorStateKey)
-
-  var shards = Set.empty[String]
   var terminating = false
   var getShardHomeRequests: Set[(ActorRef, GetShardHome)] = Set.empty
 
-  if (rememberEntities)
-    replicator ! Subscribe(AllShardsKey, self)
+  val rememberEntitiesStore =
+    if (rememberEntities) {
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1113]: `rememberEntitiesStoreProvider.map` should be enough since already an Option

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 2]
```diff
@@ -1105,36 +1106,51 @@ class DDataShardCoordinator(
...
+      if (rememberEntities) None
+      else Some(Set.empty))
 
   // This state will drop all other messages since they will be retried
-  def waitingForState(remainingKeys: Set[Key[ReplicatedData]]): Receive =
+  // Note remember entities query retries forever in store rather than here
+  def waitingForInitialState(initialState: Option[State], initialShards: Option[Set[ShardId]]): Receive =
     ({
+
+      case RememberEntitiesCoordinatorStore.RememberedShards(shardIds) =>
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1136]: I don't think waiting for `RememberedShards` is mandatory for the initialization process. Should be possible to do in the background. That can be explored in a separate PR, and maybe not important.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 3]
```diff
@@ -1236,27 +1254,6 @@ class DDataShardCoordinator(
...
-        context.become(waitingForUpdate(evt, afterUpdateCallback, newRemainingKeys))
-
-    case UpdateTimeout(AllShardsKey, Some(newShard: String)) =>
-      log.error(
-        "The ShardCoordinator was unable to update shards distributed state within 'updating-state-timeout': {} millis ({}), event={}",
-        writeMajority.timeout.toMillis,
-        if (terminating) "terminating" else "retrying",
-        evt)
-      if (terminating) {
-        context.stop(self)
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1254]: terminating not handled in same way?

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+    case Replicator.NotFound(AllShardsKey, _) =>
+      allShards = Set.empty
+      context.become(idle())
+      unstashAll()
+  }
+
+  def idle(): Receive = {
+    case RememberEntitiesCoordinatorStore.GetShards =>
+      // FIXME should we clear the set once we have replied given that we know there is only one get-request ever?
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala:79]: it's used in `AddShard`, but maybe that isn't needed?

**Review by patriknw — COMMENTED** [review]
looking great

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala [hunk 5]
```diff
@@ -27,12 +27,15 @@ import akka.persistence.SnapshotSelectionCriteria
...
-private[akka] final class EventSourcedRememberEntitiesStoreProvider(typeName: String, settings: ClusterShardingSettings)
-    extends RememberEntitiesShardStoreProvider {
+private[akka] final class EventSourcedRememberEntitiesProvider(typeName: String, settings: ClusterShardingSettings)
+    extends RememberEntitiesProvider {
 
   override def shardStoreProps(shardId: ShardId): Props =
     EventSourcedRememberEntitiesStore.props(typeName, shardId, settings)
 
+  // FIXME persistent state store deprecated but we are adding a remember entities store that is not deprecated
+  // We need a new impl for this to allow ddata + persistent remember entities
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala:37]: One challenge here is that if the event sourced actor fails (persist failure) it must still continue retrying so that AddShard or GetShards are not stuck forever.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala (hunk 2 above)
Comment (johanandren) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1136]: Keeping current impl for now, created https://github.com/akka/akka/issues/28925 to track

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala (hunk 3 above)
Comment (johanandren) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1254]: Good point. The timeout/retry was pushed into the remember entities provider and I didn't add a separate timeout here yet, need to think about whether that makes sense or not (and why waiting for this timeout before stopping makes sense).

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala (hunk 4 above)
Comment (johanandren) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala:79]: Yeah, I think we could potentially skip that, not sure what scenario that would happen and it is not harmful to add it twice.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala (hunk 5 above)
Comment (johanandren) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala:37]: Hmm, so perhaps better to move the retrying to the coordinator after all.

Would be nice to align how failure works with the remember entity store for the shard, so maybe not let-it-crash there either.

Comment (johanandren) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala:37]: The reason I pushed it into the store is that what makes sense with retrying, timeouts etc, kind of depends a bit on the store itself.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala (hunk 3 above)
Comment (johanandren) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1254]: Do'h, I'm lying I did re-add it with the `RememberEntityTimeout`

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala (hunk 5 above)
Comment (patriknw) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala:37]: Otherwise the retry logic would be in a parent supervisor of the event sourced store. It can keep current command around and retry.

Comment (johanandren) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala:38]: I updated the protocol with a fail + retrying on timeout because I think that is more clear.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+      allShards = g.get(AllShardsKey).elements
+      context.become(idle())
+      unstashAll()
+
+    case Replicator.GetFailure(AllShardsKey, _) =>
+      log.error(
+        "The ShardCoordinator was unable to get all shards state within 'waiting-for-state-timeout': {} millis (retrying)",
+        readMajority.timeout.toMillis)
+      // repeat until GetSuccess
+      context.stop(self)
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala:69]: is the repeat handled elsewhere now? update comment

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 7]
```diff
@@ -1277,10 +1285,58 @@ class DDataShardCoordinator(
...
+        settings.tuningParameters.updatingStateTimeout.toMillis,
+        if (terminating) "terminating" else "retrying",
+        evt)
+      if (terminating) context.stop(self)
+      else {
+        // retry until successful
+        rememberShardAllocated(shardId.get)
+      }
+
+      rememberShardAllocated(shard)
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1335]: rememberShardAllocated called twice

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 8]
```diff
@@ -1277,10 +1285,58 @@ class DDataShardCoordinator(
...
+        "The ShardCoordinator was unable to update shards distributed state, {} ({})",
+        if (terminating) "terminating" else "retrying",
+        evt)
+      if (terminating) context.stop(self)
+      else {
+        // retry
+        rememberShardAllocated(shardId.get)
+      }
+
+    case RememberEntitiesTimeout(shard) =>
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1321]: isn't this identical to `UpdateFailed` aside from the log message? do we need both?

Comment (patriknw) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1321]: at least extract to a method that is used by both

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 9]
```diff
@@ -1353,8 +1411,26 @@ class DDataShardCoordinator(
...
+  override def receiveTerminated: Receive =
+    super.receiveTerminated.orElse {
+      case RememberEntitiesStoreStopped =>
+        onRememberEntitiesStoreStopped()
+    }
+
+  def onRememberEntitiesStoreStopped(): Unit = {
+    // rely on backoff supervision of coordinator
+    log.error("The ShardCoordinator stopping because the remember entities store stopped")
+    context.stop(self)
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1433]: I guess that is what happens also when the coordinator state write fails when using persistent mode?

Ordinary ddata write failures doesn't end up here, they are retried?

**Review by patriknw — COMMENTED** [review]
I think it's a good to let the coordinator be responsible for the retries

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala (hunk 6 above)
Comment (patriknw) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala:69]: so, it will stop the coordinator, which will be restarted after backoff

Comment (johanandren) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala:69]: Or well, having a get-failed doesn't quite make sense either, at least not for the ddata one, since load is done eagerly on store actor start, not triggered by the request. I think it should just retry the load here instead.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala (hunk 8 above)
Comment (johanandren) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1321]: Maybe not that valuable as I was thinking with the different logs no

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala (hunk 9 above)
Comment (johanandren) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1433]: Yep, a bit of ahead of time implementation here since we don't have a persistent remember entitites store that you can run together with the ddata state store mode.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        sender() ! RememberEntitiesCoordinatorStore.UpdateDone(shardId)
+      }
+  }
+
+  def addShard(newShard: ShardId, replyTo: ActorRef): Unit = {
+    replicator ! Replicator.Update(AllShardsKey, GSet.empty[String], writeMajority, Some(newShard))(_ + newShard)
+    context.become(waitingForWrite(replyTo))
+  }
+
+  def waitingForWrite(replyTo: ActorRef): Receive = {
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala:98]: AddShard not handled here. I think we can simplify and make it more robust by tunneling the `sender()` in the `Replicator.Update.request field. Then it will come back in UpdateSuccess/UpdateTimeout and we don't need to keep track of it in the state of the actor.

Since retries are handled by the coordinator that will simplify things here. We don't need the `idle` and `waitingForWrite` behaviors. It can be one `initializing` where it's waiting for the initial GetSuccess and then one `active` that handles everything after that.

**Review by patriknw — COMMENTED** [review]
some more suggestions to DDataRememberEntitiesCoordinatorStore

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala (hunk 10 above)
Comment (johanandren) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala]: I took it one step further to a single receive, much cleaner and easier to read 👍

**Comment by johanandren** [comment]
I still see akka.cluster.sharding.DDataClusterShardingFailure failing now and then, but only when I run all tests, haven't been able to repeat it in a isolated run.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        "The remember entities store was unable to add shard [{}] (key [{}], failed with error: {})",
+        shardId,
+        key,
+        error)
+      replyTo ! RememberEntitiesCoordinatorStore.UpdateFailed(shardId)
+  }
+
+  def onGotAllShards(shardIds: Set[ShardId]): Unit = {
+    allShards = Some(shardIds)
+    coordinatorWaitingForShards match {
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala:106]: `coordinatorWaitingForShards.foreach`

**Review by patriknw — COMMENTED** [review]
excellent DDataRememberEntitiesCoordinatorStore

**Comment by johanandren** [comment]
Looks like the multi jvm test that fails now and then always has a coordinator write to LevelDB timing out, not sure it is related to this refactoring:

```
info] [JVM-2] [ERROR] [04/17/2020 14:02:56.893] [MultiNodeClusterShardingSpec-akka.actor.default-dispatcher-28] [akka://MultiNodeClusterShardingSpec@localhost:65464/system/sharding/EntityCoordinator/singleton/coordinator] Failed to persist event type [akka.cluster.sharding.ShardCoordinator$Internal$ShardHomeAllocated] with sequence number [5] for persistenceId [/sharding/EntityCoordinator].
[info] [JVM-2] akka.pattern.AskTimeoutException: Ask timed out on [Actor[akka://MultiNodeClusterShardingSpec@localhost:65462/user/store#764290164]] after [5000 ms]. Message of type [akka.persistence.journal.AsyncWriteTarget$WriteMessages] was sent by [Actor[akka://MultiNodeClusterShardingSpec/system/akka.persistence.journal.leveldb-shared#-1839094970]]. A typical reason for `AskTimeoutException` is that the recipient actor didn't send a reply.
[info] [JVM-2] 	at akka.pattern.PromiseActorRef$.$anonfun$defaultOnTimeout$1(AskSupport.scala:646)
[info] [JVM-2] 	at akka.pattern.PromiseActorRef$.$anonfun$apply$1(AskSupport.scala:667)
[info] [JVM-2] 	at akka.actor.Scheduler$$anon$7.run(Scheduler.scala:476)
[info] [JVM-2] 	at akka.dispatch.internal.SameThreadExecutionContext$$anon$1.unbatchedExecute(SameThreadExecutionContext.scala:21)
[info] [JVM-2] 	at akka.dispatch.BatchingExecutor.execute(BatchingExecutor.scala:125)
[info] [JVM-2] 	at akka.dispatch.BatchingExecutor.execute$(BatchingExecutor.scala:116)
[info] [JVM-2] 	at akka.dispatch.internal.SameThreadExecutionContext$$anon$1.execute(SameThreadExecutionContext.scala:20)
[info] [JVM-2] 	at akka.actor.LightArrayRevolverScheduler$TaskHolder.executeTask(LightArrayRevolverScheduler.scala:355)
[info] [JVM-2] 	at akka.actor.LightArrayRevolverScheduler$$anon$3.executeBucket$1(LightArrayRevolverScheduler.scala:306)
[info] [JVM-2] 	at akka.actor.LightArrayRevolverScheduler$$anon$3.nextTick(LightArrayRevolverScheduler.scala:310)
[info] [JVM-2] 	at akka.actor.LightArrayRevolverScheduler$$anon$3.run(LightArrayRevolverScheduler.scala:262)
[info] [JVM-2] 	at java.base/java.lang.Thread.run(Thread.java:834)
[info] [JVM-2]
```

**Comment by chbatey** [comment]
Might be worth running all the multi jvm/node tests on repeat on this branch

**Comment by johanandren** [comment]
Yep, definitely

**Comment by johanandren** [comment]
It could be that is something I broke with retrying failed writes of the state when refactoring.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 12]
```diff
@@ -1105,36 +1120,45 @@ class DDataShardCoordinator(
...
-  override def receive: Receive = waitingForState(allKeys)
+  override def receive: Receive =
+    waitingForInitialState(
+      None,
+      if (rememberEntities) None
+      else Some(Set.empty))
 
   // This state will drop all other messages since they will be retried
-  def waitingForState(remainingKeys: Set[Key[ReplicatedData]]): Receive =
+  // Note remember entities query retries forever in store rather than here
```
Comment (chbatey) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1149]: remembered

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 13]
```diff
@@ -1353,8 +1401,39 @@ class DDataShardCoordinator(
...
 
-  def sendAllShardsUpdate(newShard: String) = {
-    replicator ! Update(AllShardsKey, GSet.empty[String], writeMajority, Some(newShard))(_ + newShard)
+  def rememberShardAllocated(newShard: String) = {
+    log.debug("Remembering shard allocation [{}]", newShard)
+    rememberEntitiesStore.foreach(_ ! RememberEntitiesCoordinatorStore.AddShard(newShard))
+    timers.startSingleTimer(
+      RememberEntitiesTimeoutKey,
+      RememberEntitiesTimeout(newShard),
+      // FIXME more reasonable timeout here?
```
Comment (chbatey) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1410]: i think just use this for now and put a comment in reference that it is

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 14]
```diff
@@ -1277,10 +1291,38 @@ class DDataShardCoordinator(
...
     case ShardCoordinator.Internal.Terminate =>
-      log.debug("Received termination message while waiting for update")
+      log.debug("The ShardCoordinator received termination message while waiting for update")
       terminating = true
       stash()
 
+    case RememberEntitiesCoordinatorStore.UpdateDone(shard) =>
+      require(shardId.contains(shard))
+      if (!waitingForStateWrite) {
+        log.debug("The ShardCoordinator saw remember shard start successfuly written {}", evt)
```
Comment (chbatey) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1301]: fully

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 15]
```diff
@@ -1277,10 +1291,38 @@ class DDataShardCoordinator(
...
+        if (shardId.isDefined) timers.cancel(RememberEntitiesTimeoutKey)
+        unbecomeAfterUpdate(evt, afterUpdateCallback)
+      } else {
+        log.debug("The ShardCoordinator saw remember shard start successfuly written {}, waiting for state update", evt)
```
Comment (chbatey) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1305]: fully

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 16]
```diff
@@ -1206,21 +1227,35 @@ class DDataShardCoordinator(
       log.debug("Received termination message while waiting for state initialized")
       context.stop(self)
 
+    case RememberEntitiesStoreStopped =>
+      onRememberEntitiesStoreStopped()
+
     case _ => stash()
   }
 
   // this state will stash all messages until it receives UpdateSuccess
```
Comment (chbatey) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:1236]: maybe update this comment to reflect we could be waiting for 2 things

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+private[akka] object DDataRememberEntitiesCoordinatorStore {
+  def props(typeName: String, settings: ClusterShardingSettings, replicator: ActorRef, majorityMinCap: Int): Props =
+    Props(new DDataRememberEntitiesCoordinatorStore(typeName, settings, replicator, majorityMinCap))
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final class DDataRememberEntitiesCoordinatorStore(
```
Comment (chbatey) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala:30]: Looks good

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+      replyTo ! RememberEntitiesCoordinatorStore.UpdateFailed(shardId)
+
+    case Replicator.ModifyFailure(key, error, cause, Some((replyTo: ActorRef, shardId: ShardId))) =>
+      log.error(
+        cause,
+        "The remember entities store was unable to add shard [{}] (key [{}], failed with error: {})",
+        shardId,
+        key,
+        error)
+      replyTo ! RememberEntitiesCoordinatorStore.UpdateFailed(shardId)
```
Comment (chbatey) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala:99]: A bit odd we haven't different retry mechanisms for the two updates but i understand why

Comment (patriknw) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala:99]: `ModifyFailure` isn't something that will happen unless we have a bug. It's when the Update function throws.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala [hunk 19]
```diff
@@ -27,12 +27,17 @@ import akka.persistence.SnapshotSelectionCriteria
...
+    extends RememberEntitiesProvider {
 
+  // this is backed by an actor using the same events as for state-store-mode=persistence
   override def shardStoreProps(shardId: ShardId): Props =
     EventSourcedRememberEntitiesStore.props(typeName, shardId, settings)
 
+  // FIXME persistent state store deprecated but we are adding a remember entities store that is not deprecated
+  // We need a new impl for this to allow ddata + persistent remember entities
+  // For now it is anyways not possible to configure state store and remember entities store separately so this is never used
+  override def coordinatorStoreProps(): Props = ???
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala:40]: what does this mean compared to master? what happens now when mode=persistence && rememberEntities=true ?

**Review by patriknw — APPROVED** [review]
LGTM, with a question and maybe some fixme that should be turned into tickets

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala (hunk 19 above)
Comment (johanandren) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala:40]: `mode=persistence` means the persistent actor subclass of coordinator is used, that impl is left as is and uses its own state rather than an external store for remember entities, so this is never called.

**Review by chbatey — APPROVED** [review]
LGTM

**Comment by patriknw** [comment]
Merging to dev branch. The test failures are probably not related to these changes since it looked like a problem in the Shard.
